requires.py

```diff
@@ -1,31 +1,94 @@
+        comparisons to detect changes.
+        If `ca_file` is given, it will be used to replace the value of any
+        `ca_file` fields in the job.
+        If `cert_file` and `key_file` are given, they will be used to replace
+        the value of any `cert_file` and `key_file` fields in the job.
+        The charm should ensure that the request's `ca_cert`, `client_cert`
+        and `client_key` data is writen to those paths prior to calling this
+        method.
+        """
+        job_data = deepcopy(self.job_data)  # make a copy we can modify
+        job_data['job_name'] = '{}-{}'.format(self.job_name, self.request_id)
+
+        if ca_file:
+            for key, value in job_data.items():
+                # update the cert path at the job level
+                if key == 'tls_config':
+                    value['ca_file'] = str(ca_file)
+
+                # update the cert path at the SD config level
+                if key.endswith('_sd_configs'):
+                    for sd_config in value:
+                        sd_tls_config = sd_config.get('tls_config', {})
+                        if not sd_tls_config:
+                            continue
+                        if 'ca_file' in sd_tls_config:
+                            sd_tls_config['ca_file'] = str(ca_file)
+
+        if cert_file and key_file:
+            for key, value in job_data.items():
+                # update the cert/key paths at the job level
+                if key == 'tls_config':
+                    value['cert_file'] = str(cert_file)
+                    value['key_file'] = str(key_file)
+
+                # update the cert/key paths at the SD config level
+                if key.endswith('_sd_configs'):
+                    for sd_config in value:
+                        sd_tls_config = sd_config.get('tls_config', {})
+                        if not sd_tls_config:
+                            continue
+                        if 'client_file' in sd_tls_config:
+                            sd_tls_config['cert_file'] = str(cert_file)
+                        elif 'key_file' in sd_tls_config:
+                            sd_tls_config['key_file'] = str(key_file)
+
+        return json.dumps(job_data, sort_keys=True)
+
+
+class PrometheusManualRequires(Endpoint):
 
     def manage_flags(self):
         super().manage_flags()
         toggle_flag(self.expand_name('endpoint.{endpoint_name}.has_jobs'),
                     self.is_joined and self.jobs)
         toggle_flag(self.expand_name('endpoint.{endpoint_name}.new_jobs'),
-                    self.is_joined and self.new_jobs)
+                    self.is_joined and self.jobs)
 
     @property
     def jobs(self):
         """
-        Return a list of all jobs to be registered.
+        Return a list of all jobs to be registered by using application
+        relation data
         """
-        return self.all_requests
+        ret = []
+        for job in self.relations[0].received_app.values():
```

Comment:
My concern with this change is as follows:
1) Changing prometheus requires to *only* pay attention to app values means that it breaks compatibility with any charm that hasn't been updated yet.
2) Setting the request Job's request ID to a uuid4() at this point means that every time this gets reevaluated then all jobs get new random UUIDs (if I'm reading it correctly).
In which case, why would we want the uuid at all?